src/kem/kyber/kem_kyber_512.c

```diff
@@ -46,7 +46,45 @@ extern int PQCLEAN_KYBER512_AARCH64_crypto_kem_enc(uint8_t *ct, uint8_t *ss, con
 extern int PQCLEAN_KYBER512_AARCH64_crypto_kem_dec(uint8_t *ss, const uint8_t *ct, const uint8_t *sk);
 #endif
 
+#if defined(OQS_ENABLE_LIBJADE_KEM_kyber_512)
+extern int libjade_kyber512_ref_keypair(uint8_t *pk, uint8_t *sk);
+extern int libjade_kyber512_ref_enc(uint8_t *ct, uint8_t *ss, const uint8_t *pk);
+extern int libjade_kyber512_ref_dec(uint8_t *ss, const uint8_t *ct, const uint8_t *sk);
+#endif
+
+#if defined(OQS_ENABLE_LIBJADE_KEM_kyber_512_avx2)
+extern int libjade_kyber512_avx2_keypair(uint8_t *pk, uint8_t *sk);
+extern int libjade_kyber512_avx2_enc(uint8_t *ct, uint8_t *ss, const uint8_t *pk);
+extern int libjade_kyber512_avx2_dec(uint8_t *ss, const uint8_t *ct, const uint8_t *sk);
+#endif
+
+
 OQS_API OQS_STATUS OQS_KEM_kyber_512_keypair(uint8_t *public_key, uint8_t *secret_key) {
+#if defined(OQS_LIBJADE_BUILD) && (defined(OQS_ENABLE_LIBJADE_KEM_kyber_512))
+#if defined(OQS_ENABLE_LIBJADE_KEM_kyber_512_avx2)
+#if defined(OQS_DIST_BUILD)
+	if (OQS_CPU_has_extension(OQS_CPU_EXT_AVX2) && OQS_CPU_has_extension(OQS_CPU_EXT_BMI2) && OQS_CPU_has_extension(OQS_CPU_EXT_POPCNT)) {
+#endif /* OQS_DIST_BUILD */
+		return (OQS_STATUS) libjade_kyber512_avx2_keypair(public_key, secret_key);
+#if defined(OQS_DIST_BUILD)
+	} else {
+		return (OQS_STATUS) libjade_kyber512_ref_keypair(public_key, secret_key);
+	}
+#endif /* OQS_DIST_BUILD */
+#elif defined(OQS_ENABLE_KEM_kyber_512_aarch64)
+#if defined(OQS_DIST_BUILD)
+	if (OQS_CPU_has_extension(OQS_CPU_EXT_ARM_NEON)) {
+#endif /* OQS_DIST_BUILD */
+		return (OQS_STATUS) PQCLEAN_KYBER512_AARCH64_crypto_kem_keypair(public_key, secret_key);
```

Comment:
It seems to me that if the user has specified a libjade build on an ARM platform, they probably want a formally verified implementation. Perhaps we ought to use the libjade "ref" implementation irrespective of available extensions.

Reply:
At the moment libjade only supports x86. This was a poorly thought out idea for the eventuality that libjade adds selective ARM support for algorithms and I've removed this line. This "solution" clearly wouldn't work, either the code wouldn't compile or that return statement would be unreachable.

Reply:
I've edited `CMakelists.txt` so that CMake explicitly disallows building liboqs with libjade on on non-x86 platforms, I'd missed this earlier. This would need to be reevaluated once libjade supports other platforms.

Reply:
See [this comment](https://github.com/open-quantum-safe/liboqs/pull/1745#discussion_r1580123937).